File: lib/features/reminders/domain/reminder_policy.dart
import 'reminder_models.dart';

class ReminderPolicySettings {
  const ReminderPolicySettings({
    required this.enabledOffsetsDays,
    required this.lateDays,
    required this.hour,
    required this.minute,
    required this.privacy,
    required this.quietStartHour,
    required this.quietEndHour,
  });

  final Set<int> enabledOffsetsDays;
  final int lateDays;
  final int hour;
  final int minute;
  final NotificationPrivacy privacy;
  final int quietStartHour;
  final int quietEndHour;
}

class ReminderPolicy {
  const ReminderPolicy();

  List<ReminderPlan> periodPlans({
    required DateTime predictedDate,
    required String sourcePredictionId,
    required ReminderPolicySettings settings,
    required DateTime now,
  }) {
    final plans = <ReminderPlan>[];
    final offsets = settings.enabledOffsetsDays.toList()..sort((a, b) => b.compareTo(a));
    for (final offset in offsets) {
      final kind = switch (offset) {
        7 => ReminderKind.periodSevenDays,
        3 => ReminderKind.periodThreeDays,
        1 => ReminderKind.periodOneDay,
        0 => ReminderKind.periodExpectedDay,
        _ => null,
      };
      if (kind == null) continue;
      final day = DateTime(predictedDate.year, predictedDate.month, predictedDate.day)
          .subtract(Duration(days: offset));
      final target = _outsideQuietHours(
        DateTime(day.year, day.month, day.day, settings.hour, settings.minute),
        settings,
      );
      if (!target.isAfter(now)) continue;
      plans.add(
        ReminderPlan(
          id: 'period-${kind.name}',
          kind: kind,
          targetLocal: target,
          privacy: settings.privacy,
          sourcePredictionId: sourcePredictionId,
        ),
      );
    }

    if (settings.lateDays > 0) {
      final lateDay = DateTime(predictedDate.year, predictedDate.month, predictedDate.day)
          .add(Duration(days: settings.lateDays));
      final target = _outsideQuietHours(
        DateTime(lateDay.year, lateDay.month, lateDay.day, settings.hour, settings.minute),
        settings,
      );
      if (target.isAfter(now)) {
        plans.add(
          ReminderPlan(
            id: 'period-${ReminderKind.periodLate.name}',
            kind: ReminderKind.periodLate,
            targetLocal: target,
            privacy: settings.privacy,
            sourcePredictionId: sourcePredictionId,
          ),
        );
      }
    }
    return List.unmodifiable(plans);
  }

  DateTime _outsideQuietHours(
    DateTime candidate,
    ReminderPolicySettings settings,
  ) {
    final hour = candidate.hour;
    final wrapsMidnight = settings.quietStartHour > settings.quietEndHour;
    final quiet = wrapsMidnight
        ? hour >= settings.quietStartHour || hour < settings.quietEndHour
        : hour >= settings.quietStartHour && hour < settings.quietEndHour;
    if (!quiet) return candidate;

    // Preserve the intended reminder calendar day. This avoids a late-night
    // reminder slipping into the following cycle-relative day.
    return DateTime(
      candidate.year,
      candidate.month,
      candidate.day,
      settings.quietEndHour,
    );
  }
}

File: web/src/domain/reminders/reminder-policy.ts
export type NotificationPrivacy = 'maximum' | 'balanced' | 'detailed';

export type PeriodReminderKind =
  | 'periodSevenDays'
  | 'periodThreeDays'
  | 'periodOneDay'
  | 'periodExpectedDay'
  | 'periodLate';

export type ReminderPolicySettings = {
  enabledOffsetsDays: readonly number[];
  lateDays: number;
  hour: number;
  minute: number;
  privacy: string;
  quietStartHour: number;
  quietEndHour: number;
};

export type ReminderPlan = {
  kind: PeriodReminderKind;
  targetLocal: string;
  privacy: NotificationPrivacy;
};

export type PeriodReminderInput = {
  predictedDate: string;
  sourcePredictionId: string;
  now: string;
  settings: ReminderPolicySettings;
};

const DAY_MS = 86_400_000;

function parseDateOnly(value: string): number {
  const match = /^(\d{4})-(\d{2})-(\d{2})$/.exec(value);
  if (!match) throw new Error(`reminder-v1 requires YYYY-MM-DD: ${value}`);
  const millis = Date.UTC(Number(match[1]), Number(match[2]) - 1, Number(match[3]));
  if (new Date(millis).toISOString().slice(0, 10) !== value) {
    throw new Error(`reminder-v1 received an invalid calendar date: ${value}`);
  }
  return millis;
}

function addDays(value: string, days: number): string {
  return new Date(parseDateOnly(value) + days * DAY_MS).toISOString().slice(0, 10);
}

function twoDigits(value: number): string {
  return String(value).padStart(2, '0');
}

function privacy(value: string): NotificationPrivacy {
  if (value === 'maximum' || value === 'balanced' || value === 'detailed') return value;
  throw new Error(`unsupported notification privacy level: ${value}`);
}

function kindForOffset(offset: number): PeriodReminderKind | null {
  switch (offset) {
    case 7: return 'periodSevenDays';
    case 3: return 'periodThreeDays';
    case 1: return 'periodOneDay';
    case 0: return 'periodExpectedDay';
    default: return null;
  }
}

function quietAdjustedTime(settings: ReminderPolicySettings): { hour: number; minute: number } {
  const { hour, minute, quietStartHour, quietEndHour } = settings;
  const wrapsMidnight = quietStartHour > quietEndHour;
  const isQuiet = wrapsMidnight
    ? hour >= quietStartHour || hour < quietEndHour
    : hour >= quietStartHour && hour < quietEndHour;

  return isQuiet ? { hour: quietEndHour, minute: 0 } : { hour, minute };
}

function targetLocal(date: string, settings: ReminderPolicySettings): string {
  const time = quietAdjustedTime(settings);
  return `${date}T${twoDigits(time.hour)}:${twoDigits(time.minute)}`;
}

function wallClockMillis(value: string): number {
  const normalized = value.length === 16 ? `${value}:00` : value;
  const millis = Date.parse(`${normalized}Z`);
  if (!Number.isFinite(millis)) throw new Error(`invalid local wall-clock value: ${value}`);
  return millis;
}

export function planPeriodReminders(input: PeriodReminderInput): ReminderPlan[] {
  // sourcePredictionId is deliberately accepted at the policy boundary so callers
  // can tie regenerated schedules to a prediction generation. The frozen shared
  // vector shape contains only the delivery plan itself.
  if (input.sourcePredictionId.length === 0) throw new Error('sourcePredictionId is required');

  const settings = input.settings;
  const output: ReminderPlan[] = [];
  const offsets = [...settings.enabledOffsetsDays].sort((left, right) => right - left);

  for (const offset of offsets) {
    const kind = kindForOffset(offset);
    if (!kind) continue;
    const candidate = targetLocal(addDays(input.predictedDate, -offset), settings);
    if (wallClockMillis(candidate) <= wallClockMillis(input.now)) continue;
    output.push({ kind, targetLocal: candidate, privacy: privacy(settings.privacy) });
  }

  if (settings.lateDays > 0) {
    const candidate = targetLocal(addDays(input.predictedDate, settings.lateDays), settings);
    if (wallClockMillis(candidate) > wallClockMillis(input.now)) {
      output.push({
        kind: 'periodLate',
        targetLocal: candidate,
        privacy: privacy(settings.privacy),
      });
    }
  }

  return output;
}

export function notificationBody(daysBefore: number, level: string): string {
  const selected = privacy(level);
  if (selected === 'maximum') return 'You have a reminder.';
  if (selected === 'balanced') return 'Your cycle reminder is ready.';

  switch (daysBefore) {
    case 7: return 'Your period may begin in about 7 days.';
    case 3: return 'Your period may begin in about 3 days.';
    case 1: return 'Your period may begin tomorrow.';
    case 0: return 'Your period is expected around today.';
    default: return 'Your health reminder is ready.';
  }
}
